packages/workspace/src/browser/workspace-trust-service.ts

```diff
@@ -0,0 +1,115 @@
+    protected readonly preferences: PreferenceService;
+
+    @inject(StorageService)
+    protected readonly storage: StorageService;
+
+    @inject(MessageService)
+    protected readonly messageService: MessageService;
+
+    @inject(WorkspaceTrustPreferences)
+    protected readonly workspaceTrustPref: WorkspaceTrustPreferences;
+
+    protected trusted = new Deferred<boolean>();
+    protected _isTrusted: boolean | undefined;
+
+    @postConstruct()
+    protected async init(): Promise<void> {
+        await this.workspaceService.ready;
+        this.trusted.promise.then(trusted => this._isTrusted = trusted);
+        const implicitTrust = await this.calculateWorkspaceTrust();
+        if (implicitTrust !== undefined) {
+            this.trusted.resolve(implicitTrust);
+        }
+    }
+
+    get isTrusted(): boolean | undefined {
+        return this._isTrusted;
+    }
+
+    get onTrusted(): Promise<boolean> {
+        return this.trusted.promise;
+    }
+
+    protected async calculateWorkspaceTrust(): Promise<boolean | undefined> {
+        await this.preferences.ready;
+        const autoTrust = FrontendApplicationConfigProvider.get()[CONFIG_AUTO_TRUST];
+        if (typeof autoTrust === 'boolean') {
+            return autoTrust;
+        }
+
+        if (!this.workspaceTrustPref[WORKSPACE_TRUST_ENABLED]) {
+            // in VS Code if workspace trust is disabled, we implicitly trust the workspace
+            return true;
+        }
+
+        if (this.workspaceTrustPref[WORKSPACE_TRUST_EMPTY_WINDOW] && !this.workspaceService.workspace) {
+            return true;
+        }
+
+        const trustPrompt = this.workspaceTrustPref[WORKSPACE_TRUST_STARTUP_PROMPT];
+        if (trustPrompt === WORKSPACE_TRUST_STARTUP_PROMPT_ONCE) {
+            const storedTrust = await this.storage.getData<boolean>(STORAGE_TRUSTED);
+            return storedTrust;
+        }
+        // reset stored trusted setting
+        this.storage.setData(STORAGE_TRUSTED, undefined);
+        if (!trustPrompt || trustPrompt === WORKSPACE_TRUST_STARTUP_PROMPT_NEVER) {
+            return false;
+        }
+        return undefined;
+    }
+
+    async requestWorkspaceTrust(): Promise<boolean | undefined> {
+        if (this.trusted.state === 'unresolved') {
+            const queryTrusted = await this.messageService.info(nls.localize('theia/workspace/trustRequest',
+                'An extension requests workspace trust but the corresponding API is not yet fully supported. Do you want to trust this workspace?'),
+                Dialog.YES, Dialog.NO);
+            const trusted = queryTrusted === Dialog.YES;
+
+            const trustPrompt = this.workspaceTrustPref[WORKSPACE_TRUST_STARTUP_PROMPT];
+            if (trustPrompt === WORKSPACE_TRUST_STARTUP_PROMPT_ONCE) {
+                await this.storage.setData(STORAGE_TRUSTED, trusted);
+            }
+            this.trusted.resolve(trusted);
+        }
+        return this.trusted.promise;
+    }
```

Comment:
If it's reasonable to wait to run `calculateWorkspaceTrust` until the first call to `requestWorkspaceTrust`, it would be better to have the query as the final fallback if we arrive at the end of `calculateWorkspaceTrust` without having returned a `boolean` value.

Reply:
I agree but I think we need to calculate the flag early on if possible, see my previous answer.